siliconcompiler/design.py

```diff
@@ -0,0 +1,535 @@
+
+    def get_undefine(self, fileset: str) -> List[str]:
+        """Returns undefined macros for a fileset.
+
+        Args:
+           fileset (str): Fileset name.
+
+       Returns:
+           list[str]: List of macro (un)definitions
+
+        """
+        return self._get(fileset, 'undefine')
+
+    ###############################################
+    def set_libdir(self, fileset: str, value: str) -> List[str]:
+        """Sets dynamic library directories for a fileset.
+
+        Args:
+           fileset (str): Fileset name.
+           value (str or List[str]): Library directories.
+        """
+        return self._set(fileset, 'libdir', value, typelist=[str, list])
+
+    def get_libdir(self, fileset: str) -> List[str]:
+        """Returns dynamic library directories for a fileset.
+
+        Args:
+           fileset (str): Fileset name.
+
+        Returns:
+           list[str]: List of library directories.
+
+        """
+        return self._get(fileset, 'libdir')
+
+    ###############################################
+    def set_lib(self, fileset: str, value: str) -> List[str]:
+        """Sets list of dynamic libraries for a fileset.
+
+        Args:
+           fileset (str): Fileset name.
+           value (str or List[str]): Library directory names.
+
+        """
+        return self._set(fileset, 'lib', value, typelist=[str, list])
+
+    def get_lib(self, fileset: str) -> List[str]:
+        """Returns list of dynamic libraries for a fileset.
+
+        Args:
+           fileset (str): Fileset name.
+
+        Returns:
+           list[str]: List of libraries.
+
+        """
+        return self._get(fileset, 'lib')
+
+    ###############################################
+    def set_param(self, fileset: str, name: str, value: str) -> str:
+        """Sets a named design parameter for a fileset.
+
+        Args:
+            fileset (str): Fileset name.
+            name (str): Parameter name.
+            value (str): Parameter value.
+        """
+
+        if not isinstance(fileset, str):
+            raise TypeError("fileset key must be a string")
+
+        if not isinstance(value, str) or value is None:
+            raise TypeError("param value must be a string")
+
+        self.set('fileset', fileset, 'param', name, value)
+        return self.get('fileset', fileset, 'param', name)
+
+    def get_param(self, fileset: str, name: str) -> str:
+        """Returns value of a named design parameter.
+
+        Args:
+           name (str): Parameter name.
+           fileset (str): Fileset name.
+
+        Returns:
+            str: Parameter value
+        """
+        if not isinstance(fileset, str):
+            raise RuntimeError("fileset value must be a string")
+        return self.get('fileset', fileset, 'param', name)
+
+    ###############################################
+    def add_file(self, fileset: str, filename: str, filetype=None, package=None):
```

Comment:
do we want a way to clobber the files? ie. reset the list, which calls set instead of add?

Reply:
sold on adding clobber.

Reply:
done